### ISSUE TITLE
bug: Debug: Attach to Node Process sends SIGUSR1 and attaches to wrong process

### ISSUE THREAD

**Issue by neil176** [microsoft/vscode#210204]
When using "Debug: Attach to Node Process", it puts the wrong process into debug mode and attaches to it

Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.88.1
- OS Version: Darwin arm64 23.4.0

Issue started on 1.88.0 and was not present on 1.87.2

Steps to Reproduce:
What I set up to repro this is two scripts that just report their pid and then sleep in a while loop.
- Run script A
- Run script B
- Copy the pid of script A process
- Use command palette to "Debug: Attach to node process"
- Paste in pid of script A process
- VSCode sends a SIGUSR1 to script B process and attaches debugger (separately confirmed with an explicit `process.on('SIGUSR1' ...)` listener)

This seems to only reproduce when I am trying to attach to the lower numbered pid process, with the higher numbered one getting attached to incorrectly (I think I had one case of a lower number pid getting attached to, but can't repro reliably). If I separately send a `SIGUSR1` to the process running script A, and then attempt to debug it, the debugger correctly attaches to it, but also incorrectly sends a `SIGUSR1` to script B process (where it naturally complains that port 9229 is already in-use).

![Screenshot 2024-04-10 at 06 41 07](https://github.com/microsoft/vscode/assets/27479283/2beeee24-c5ab-429c-a4d5-4adf904efa4d)
![Screenshot 2024-04-10 at 06 41 24](https://github.com/microsoft/vscode/assets/27479283/1c436c27-17f2-43b9-af8e-d97fba503a24)
![Screenshot 2024-04-10 at 06 41 31](https://github.com/microsoft/vscode/assets/27479283/0903957b-9fad-4c82-a412-6aa5e25813e0)

It's not clear to me if the issue is specific to debug-related features, or maybe something like the process picker after first selecting Debug: Attach to node process in the command palette 

Most of this was copied from comments in https://github.com/microsoft/vscode/issues/209785 which was deemed a separate issue.

**Comment by wxpftd** [comment]
I get this error when i use "Debug: Attach to Node Process"，when i upgrade to version 1.88。downgrade to 1.86.2 didn't help.

Could not read source map for file:///~/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/out/vs/out-vscode/vs/loader.js: ENOENT: no such file or directory, open '~/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/out/vs/out-vscode/vs/nls.js.map'

Version: 1.86.2
Commit: 903b1e9d8990623e3d7da1df3d33db3e42d80eda
Date: 2024-02-13T19:42:12.210Z
Electron: 27.2.3
ElectronBuildId: 26908389
Chromium: 118.0.5993.159
Node.js: 18.17.1
V8: 11.8.172.18-electron.0
OS: Darwin arm64 23.4.0

**Comment by neil176** [comment]
Some notes from exploring this a bit today:
It also repros on:
```
Version: 1.88.1
Commit: e170252f762678dec6ca2cc69aba1570769a5d39
Date: 2024-04-10T17:34:12.840Z
Electron: 28.2.8
ElectronBuildId: 27744544
Chromium: 120.0.6099.291
Node.js: 18.18.2
V8: 12.0.267.19-electron.0
OS: Linux x64 6.8.0-76060800daily20240311-generic
```
I _cannot_ repro developing on vscode running in a devcontainer at current `main` or `1.88.1`. It's not clear to me if there's some nondeterminism or race condition that is less likely under this setup. I'm guessing the root cause is somewhere around  the `processPicker` over in https://github.com/microsoft/vscode-js-debug/blob/main/src/ui/processPicker.ts#L135 but I had some trouble getting set up while debugging to artificially add some waits and latency when trying to force it to repro. I'm also wondering if there's some correlation with the number of processes being scanned and added to the picker and at what rate, based on the fact that in the more limited container env with few processes I can't repro, but outside of that context on either my personal machine running a bunch of stuff or my corporate machine with some massive quantity of background processes it does.

Sorry that's not much to go on but just thought I'd follow up with as far as I was able to get on my own.

**Comment by connor4312** [comment]
Since the display is right, and also because it likes to select the adjacent item, I'm suspicious of the picker itself. In the quickpick we add items as we discover them and re-assign the array, which I probably uncommon: https://github.com/microsoft/vscode-js-debug/blob/0b40365364f2867691701bff5aedf68d7fa0fe9d/src/ui/processPicker.ts#L169-L181

@TylerLeonhardt do you already have an issue tracking this?

**Comment by TylerLeonhardt** [comment]
Can you try [VS Code Insiders](http://code.visualstudio.com/insiders/)? I believe this is already fixed there.

**Comment by neil176** [comment]
@TylerLeonhardt  Thank you! This appears fixed on insiders. Have confirmed on:
```
Version: 1.89.0-insider (Universal)
Commit: b58957e67ee1e712cebf466b995adf4c5307b2bd
Date: 2024-05-01T02:05:29.995Z
Electron: 28.2.8
ElectronBuildId: 27744544
Chromium: 120.0.6099.291
Node.js: 18.18.2
V8: 12.0.267.19-electron.0
OS: Darwin arm64 23.4.0
```

Just for my curiosity, any chance you could point me to the PR or commit that fixed it?
Thank you both!

**Comment by TylerLeonhardt** [comment]
This is what originally fixed your issue (this fix is in 1.89 stable): https://github.com/microsoft/vscode/pull/210803

But in the _latest_ insiders (1.90) has an improvement on that with this fix: https://github.com/microsoft/vscode/pull/211399
